File: day 9/main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashSet;
import java.util.Set;

class Point {
    public int x;
    public int y;

    Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public int hashCode() {
        return x * 1000 + y;
    }

    @Override
    public boolean equals(Object obj) {
        return (obj instanceof Point) && (((Point) obj).x == this.x) && (((Point) obj).y == this.y);
    }
}

class Main {
    public static Point calc_new_tail_pos(Point tail, Point head) {
        Point offset = new Point(head.x - tail.x, head.y - tail.y);
        int len = Math.abs(offset.x) + Math.abs(offset.y);
        Boolean diagonal = offset.x != 0 && offset.y != 0;
        if (len == 2 && !diagonal) {
            // move towards head
            return new Point(tail.x + offset.x / 2, tail.y + offset.y / 2);
        } else if ((len == 3 || len == 4) && diagonal) {
            // move towards diagonal pos
            return new Point(tail.x + Integer.signum(offset.x), tail.y + Integer.signum(offset.y));
        } else {
            // dont move
            return new Point(tail.x, tail.y);
        }
    }

    public static Point calc_direction(char direction) {
        if (direction == 'R') {
            return new Point(1, 0);
        } else if (direction == 'U') {
            return new Point(0, 1);
        } else if (direction == 'L') {
            return new Point(-1, 0);
        } else {
            return new Point(0, -1);
        }
    }

    public static void main(String[] args) throws IOException {
        Set<Point> visited = new HashSet<Point>();
        Point[] rope = new Point[10];
        for (int i = 0; i < rope.length; i++) {
            rope[i] = new Point(0, 0);
        }

        Files.lines(Paths.get("input")).forEach((line) -> {
            Point direction = calc_direction(line.charAt(0));
            int magnitude = Integer.parseInt(line.substring(2));
            for (int i = 1; i <= magnitude; i++) {
                rope[0] = new Point(rope[0].x + direction.x, rope[0].y + direction.y);
                for (int j = 1; j < rope.length; j++) {
                    rope[j] = calc_new_tail_pos(rope[j], rope[j - 1]);
                }
                visited.add(new Point(rope[rope.length - 1].x, rope[rope.length - 1].y));
            }
        });
        System.out.println(visited.size());
    }
}

File: day 9/src/main.kt
import java.io.File
import java.io.InputStream
import kotlin.math.*

fun calc_new_tail_pos(tail: Pair<Int, Int>, head: Pair<Int, Int>): Pair<Int, Int> {
    val offset = Pair(head.first - tail.first, head.second - tail.second);
    val len = abs(offset.first) + abs(offset.second)
    val diagonal = offset.first != 0 && offset.second != 0
    if (len == 2 && !diagonal) {
        // move towards head
        return Pair(tail.first + offset.first / 2, tail.second + offset.second / 2)
    } else if ((len == 3 || len == 4) && diagonal) {
        // move towards diagonal pos
        return Pair(tail.first + offset.first.sign, tail.second + offset.second.sign)
    } else {
        // dont move
        return tail
    }
}

fun calc_direction(direction: Char): Pair<Int, Int> {
    if (direction == 'R') {
        return Pair(1, 0)
    } else if (direction == 'U') {
        return Pair(0, 1)
    } else if (direction == 'L') {
        return Pair(-1, 0)
    } else if (direction == 'D') {
        return Pair(0, -1)
    } else {
        // should never happen
        return Pair(0, 0)
    }
}

fun main(args: Array<String>) {
    val inputStream: InputStream = File("input").inputStream()
    val lineList = mutableListOf<String>()

    var visited = mutableSetOf<Pair<Int, Int>>()
    var rope = Array(10) {Pair(0, 0)}

    inputStream.bufferedReader().forEachLine { lineList.add(it) } 

    lineList.forEach { 
        val direction = calc_direction(it[0])
        val magnitude = it.drop(2).toInt()
        for (i in 1..magnitude) {
            rope[0] = Pair(rope[0].first + direction.first, rope[0].second + direction.second)
            for (j in 1..rope.size-1) {
                rope[j] = calc_new_tail_pos(rope[j], rope[j-1])
            }
            visited += rope[rope.size-1]
        }
     }

     println(visited.size)
}